pkg/controller/nodeipam/ipam/multi_cidr_range_allocator.go

```diff
@@ -0,0 +1,1105 @@
+	// clusterCIDRConfigSynced returns true if the clustercidrconfig shared informer has been synced at least once.
+	clusterCIDRConfigSynced cache.InformerSynced
+	// Channel that is used to pass updating Nodes and their reserved CIDRs to the background.
+	// This increases a throughput of CIDR assignment by not blocking on long operations.
+	nodeCIDRUpdateChannel chan multiCIDRNodeReservedCIDRs
+	recorder              record.EventRecorder
+	// Keep a set of nodes that are currently being processed to avoid races in CIDR allocation.
+	lock              *sync.Mutex
+	nodesInProcessing sets.String
+	// queue is where incoming work is placed to de-dup and to allow "easy"
+	// rate limited requeues on errors
+	queue workqueue.RateLimitingInterface
+	// cidrMap maps ClusterCIDRConfig labels to ClusterCIDR objects.
+	cidrMap map[string][]*cidrset.ClusterCIDR
+}
+
+// NewMultiCIDRRangeAllocator returns a CIDRAllocator to allocate CIDRs for node (one for each ip family).
+// Caller must always pass in a list of existing nodes to the new allocator.
+// NodeList is only nil in testing.
+func NewMultiCIDRRangeAllocator(
+	client clientset.Interface,
+	nodeInformer informers.NodeInformer,
+	clusterCIDRConfigInformer networkinginformers.ClusterCIDRConfigInformer,
+	allocatorParams CIDRAllocatorParams,
+	nodeList *v1.NodeList,
+	cccList *networkingv1alpha1.ClusterCIDRConfigList) (CIDRAllocator, error) {
+	if client == nil {
+		klog.Fatalf("client is nil when starting CIDR Allocator")
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{
+		Component: "multiCIDRRangeAllocator",
+	})
+	eventBroadcaster.StartStructuredLogging(0)
+	klog.V(0).Infof("Sending events to api server.")
+	eventBroadcaster.StartRecordingToSink(
+		&v1core.EventSinkImpl{
+			Interface: client.CoreV1().Events(""),
+		})
+
+	if client.CoreV1().RESTClient().GetRateLimiter() != nil {
+		ratelimiter.RegisterMetricAndTrackRateLimiterUsage("multi_cidr_range_allocator", client.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	ra := &multiCIDRRangeAllocator{
+		client:                  client,
+		nodeLister:              nodeInformer.Lister(),
+		nodesSynced:             nodeInformer.Informer().HasSynced,
+		clusterCIDRConfigLister: clusterCIDRConfigInformer.Lister(),
+		clusterCIDRConfigSynced: clusterCIDRConfigInformer.Informer().HasSynced,
+		nodeCIDRUpdateChannel:   make(chan multiCIDRNodeReservedCIDRs, cidrUpdateQueueSize),
+		recorder:                recorder,
+		nodesInProcessing:       sets.NewString(),
+		queue:                   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "multi_cidr_range_allocator"),
+		lock:                    &sync.Mutex{},
+		cidrMap:                 make(map[string][]*cidrset.ClusterCIDR, 0),
+	}
+
+	// This is only set for testing purposes.
+	if len(allocatorParams.CIDRMap) > 0 {
+		ra.cidrMap = allocatorParams.CIDRMap
```

Comment:
We should log a warning here that something intended only for testing is being set.

It should be clear that if a user sees this in production, there is a misconfiguration or bug.

Reply:
Done